### ISSUE TITLE
How can I specify HTTP Method in my routes

### ISSUE THREAD

**Issue by anhar** [pointfreeco/vapor-routing#25]
Hi y'all!

I saw the fantastic video on the PointFree website demonstrating the ability of this library and I wanted to play around with it.
I've got myself a simple little Vapor app with a `House` entity and `Room` entity in order to have a one-to-many relation.

I'm setting up my router like this:
```swift
func setupRouter(app: Application) throws {
    let houseRouter = OneOf {
        Route(.case(HouseRoute.house)) {
            Path { "houses"; UUID.parser() }
        }
        Route(.case(HouseRoute.create)) {
            Path { "houses"; }
        }
    }
    app.mount(houseRouter, use: houseHandler)
}
```

And then I have my handler function looks like this:
```swift
func houseHandler(request: Request,
                  route: HouseRoute) async throws -> any AsyncResponseEncodable {
    switch route {
    case let .house(houseId: houseId):
        guard let house: House = try await House.query(on: request.db)
            .filter(\House.$id == houseId)
            .first() else {
            throw Abort(.notFound)
        }
        return HouseResponse(house: house)
    case .create:
        let house: House = try request.content.decode(House.self)
        try await house.create(on: request.db)
        return HouseResponse(house: house)
    }
}
```

When running the request with a `POST` HTTP method it fails with the following error:
<img width="626" alt="image" src="https://user-images.githubusercontent.com/800440/226489875-cda2b519-12b3-4008-819a-f8e987577f53.png">

However if I run the request with a `GET` HTTP method it succeeds even though a `GET` request shouldn't have a HTTP body according to the spec:
<img width="778" alt="image" src="https://user-images.githubusercontent.com/800440/226490110-5b8770b7-2be5-4d54-88cf-4179768b2d17.png">

Now my question is how do I specify in the `setupRouter` method that the HouseRoute.create enum case maps to a `POST` HTTP method?

**Comment by anhar** [comment]
Okay so after doing some digging I found out that the root cause of this failing is because of the [following line in Route.swift](https://github.com/pointfreeco/swift-url-routing/blob/9294f32d7f623f9665785ee3e0529d18c34a23f2/Sources/URLRouting/Route.swift#L70) in the [swift-url-routing library](https://github.com/pointfreeco/swift-url-routing).

I forked both repositories and updated the method to this which solved the issue:
```swift
@inlinable
  public func parse(_ input: inout URLRequestData) throws -> Parsers.Output {
    let output = try self.parsers.parse(&input)
    if input.method != nil {
        if input.method == "GET" {
            try Method.get.parse(&input)
        }
        else if input.method == "POST" {
            try Method.post.parse(&input)
        }
    }
    try PathEnd().parse(input)
    return output
  }
```

However I'm not sure how the underlying library works and if you want to just go ahead and hardcode the different HTTP methods in a beautiful nested if else statement 🤷‍♂️

**Comment by iampatbrown** [comment]
@anhar You can use the [Method](https://pointfreeco.github.io/swift-url-routing/main/documentation/urlrouting/method) parser:

```swift
Route(.case(HouseRoute.create)) {
  Method.post
  Path { "houses" }
}
```

> I found out that the root cause of this failing is because of the [following line in Route.swift](https://github.com/pointfreeco/swift-url-routing/blob/9294f32d7f623f9665785ee3e0529d18c34a23f2/Sources/URLRouting/Route.swift#L70)

The parser uses the `GET` method by default if it isn't set on the request.

The [swift-url-routing Documentation](https://pointfreeco.github.io/swift-url-routing/main/documentation/urlrouting/) is a good place to start to see what's available. It might be worth checking out the [benchmark suite](https://github.com/pointfreeco/swift-url-routing/blob/main/Sources/swift-url-routing-benchmark/Routing.swift) or even the [PointFreeRouter](https://github.com/pointfreeco/pointfreeco/blob/main/Sources/PointFreeRouter/Routes.swift) for some example use cases.

**Comment by anhar** [comment]
> The parser uses the GET method by default if it isn't set on the request.

That's not true though, the `URLRouting.URLRequestData` object has its `method` property set to `POST`. See the screenshot below:
![bug](https://user-images.githubusercontent.com/800440/226508718-320d56fc-5c18-43d1-b409-9a30501fcb07.jpg)

This code snippet sets the `method` property to `nil` somewhere down the callstack:
```swift
Route(.case(HouseRoute.create)) {
    Method.post
    Path { "houses"; }
}
```

See screenshot here:
![bug2](https://user-images.githubusercontent.com/800440/226509241-9d383f19-41f7-4bbc-a1e4-c6242060df8c.jpg)

**Comment by iampatbrown** [comment]
@anhar You're right. I meant the parser tries to parse the `GET` method by default. That way `Method.get` can be omitted for `GET` (and `HEAD`) requests.
The `Method` parser, will parse the `method` from the request data which sets it to `nil`.
Does using `Method.post` solve the problem or is it still not working as expected?

**Comment by mbrandonw** [comment]
Hi @anhar, Pat is correct in this situation, and your PR is not how the library should work. It may seem weird, but [this logic](https://github.com/pointfreeco/swift-url-routing/blob/9294f32d7f623f9665785ee3e0529d18c34a23f2/Sources/URLRouting/Route.swift#L68-L75) in `Route`'s parse method is correct:

```swift
  public func parse(_ input: inout URLRequestData) throws -> Parsers.Output {
    let output = try self.parsers.parse(&input)
    if input.method != nil {
      try Method.get.parse(&input)
    }
    try PathEnd().parse(input)
    return output
  }
```

In this method, `input` is the incoming URL request to your server. If after running the parser on this request the `method` is still not-`nil`, then it means that your parser chose not to parse the HTTP method ever (this is because the `Method` parser [consumes](https://github.com/pointfreeco/swift-url-routing/blob/9294f32d7f623f9665785ee3e0529d18c34a23f2/Sources/URLRouting/Method.swift#L49-L50) the `method` if it succeeds). So, if that is the case then we _force_ the method to be a GET or HEAD.

So, in your first screenshot it is completely expected that `input.method` would be POST because your parser did not try parsing the method. And in your second screenshot it is correct for `input.method` to be `nil` because then you did start using `Method.post` to parse the method.

With that said, I do believe you want to specify `Method.post` in your router for creating a house:

```swift
let houseRouter = OneOf {
  Route(.case(HouseRoute.house)) {
    Path { "houses"; UUID.parser() }
  }
  Route(.case(HouseRoute.create)) {
    Method.post
    Path { "houses"; }
  }
}
```

You can also specify `Method.get` in the first route, but it is not necessary because it is assumed when left off.
